src/analyses/base.ml

```diff
@@ -1997,6 +1997,20 @@ struct
     let st' = invalidate ~deep:false ~ctx (Analyses.ask_of_ctx ctx) gs st shallow_addrs in
     invalidate ~deep:true ~ctx (Analyses.ask_of_ctx ctx) gs st' deep_addrs
 
+  let check_free_of_non_heap_mem ctx special_fn ptr =
+    match ctx.ask (Queries.MayPointTo ptr) with
+    | a when not (Queries.LS.is_top a) ->
+      let warn_if_not_heap_var special_fn var =
+        if not (ctx.ask (Queries.IsHeapVar var)) then
+          M.warn ~category:(Behavior (Undefined InvalidMemoryDeallocation)) ~tags:[CWE 590] "Free of non-dynamically allocated memory in function %s for pointer %a" special_fn.vname d_exp ptr
+      in
```

Comment:
Does it suffice to emit one warning if such an element exists instead of calling `warn` for every var? For large points-to sets, this may be a bit more efficient.

Reply:
Good point. I changed this to now warn once for the entire points-to set in case there exists a var that is not a heap var (cf. https://github.com/goblint/analyzer/pull/1099/commits/acb1ca83d00ae5f09167e1b30b775a465736ecd1)